doc/architectural/CBMC-developer-guide.md

```diff
@@ -0,0 +1,507 @@
+first encountered, and the same index is used for all instances. 
+
+See documentation at \ref dstringt.
+
+## irept: a 4-triple (data, named-sub, comments, sub) ##
+See documentation at \ref irept.
+
+As that documentation says, `irept`s are generic tree nodes. You should
+think of them as having a single string (data, actually an `irep_idt`) and
+lots of child nodes, some of which are numbered (sub) and some of which are
+labelled, and the label can either start with a “#” (comments-sub) or without
+one (named-sub). The meaning of the “#” is that this child should not be
+considered important, for example it shouldn’t be counted when comparing two
+`irept`s for equality.
+
+## typet ##
+
+To be documented.
+
+### symbol_typet ###
+
+To be documented.
+
+## exprt ##
+\ref exprt is the class to represent an expression. It inherits from \ref irept,
+and the only things it adds to it are that every \ref exprt has a named sub
+containing its type and everything in the sub of an \ref exprt is again an
+\ref exprt, not just an \ref irept. You can think of \ref exprt as a node in the
+abstract syntax tree for an expression. There are many classes that
+inherit from \ref exprt and which represent more specific things. For
+example, there is \ref minus_exprt, which has a sub of size 2 (for the two
+argument of minus).
+
+Recall that every \ref irept has one piece of data of its own, i.e. its
+`id()`, and all other information is in its `named_sub`, `comments` or
+`sub`. For `exprt`s, the `id()` is used to specify why kind of \ref exprt this is,
+so a \ref minus_exprt has `ID_minus` as its `id()`. This means that a
+\ref minus_exprt can be passed wherever an \ref exprt is expected, and if you
+want to check if the expression you are looking at is a minus
+expression then you have to check its `id()` (or use
+`can_cast_expr<minus_exprt>`).
+
+## codet ##
+\ref exprt represents expressions and \ref codet represents statements. \ref codet
+inherits from \ref exprt, so all `codet`s are `exprt`s, with `id()` `ID_code`.
+Many different kinds of statements inherit from \ref codet, and they are
+distinguished by their `statement()`. For example, a while loop would be
+represented by a \ref code_whilet, which has `statement()` `ID_while`.
+\ref code_whilet has two operands in its sub, and helper functions to make
+it easier to use: `cond()` returns the first subexpression, which
+is the condition for the while loop, and `body()` returns the second
+subexpression, which is the body of the while loop - this has to be
+a \ref codet, because the body of a while loop is a statement.
+
+## symbolt, symbol_table, and namespacet ##
+
+To be documented.
+
+### Symbol lifetimes, symbol modes, name, base-name, pretty-name; semantic of lifetimes for symex? ###
+
+To be documented.
+
+### Storing symbols and hiding symbols (namespacet) ###
+
+To be documented.
+
+### ns.follow ##
+
+To be documented.
+
+
+## Examples: how to represent the AST of statements, focus on java ##
```

Comment:
Why focus on Java?

Reply:
I've changed it to "how to represent the AST of statements, in C and in java"